pkg/controllers/updaterun/validation.go

```diff
@@ -0,0 +1,319 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package updaterun
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+
+	placementv1alpha1 "go.goms.io/fleet/apis/placement/v1alpha1"
+	placementv1beta1 "go.goms.io/fleet/apis/placement/v1beta1"
+	"go.goms.io/fleet/pkg/utils/condition"
+	"go.goms.io/fleet/pkg/utils/controller"
+)
+
+// validate validates the clusterStagedUpdateRun status and ensures the update can be continued.
+// The function returns the index of the stage that is updating, and the list of clusters that are scheduled to be deleted.
+// If the updating stage index is -1, it means all stages are finished, and the clusterStageUpdateRun should be marked as finished.
+// If the updating stage index is 0, the next stage to be updated is the first stage.
+// If the updating stage index is len(updateRun.Status.StagesStatus), the next stage to be updated will be the delete stage.
+func (r *Reconciler) validate(
+	ctx context.Context,
+	updateRun *placementv1alpha1.ClusterStagedUpdateRun,
+) (int, []*placementv1beta1.ClusterResourceBinding, []*placementv1beta1.ClusterResourceBinding, error) {
+	// Some of the validating function changes the object, so we need to make a copy of the object.
+	updateRunRef := klog.KObj(updateRun)
+	updateRunCopy := updateRun.DeepCopy()
+	klog.V(2).InfoS("Start to validate the clusterStagedUpdateRun", "clusterStagedUpdateRun", updateRunRef)
+
+	// Validate the ClusterResourcePlacement object referenced by the ClusterStagedUpdateRun.
+	placementName, err := r.validateCRP(ctx, updateRunCopy)
+	if err != nil {
+		return -1, nil, nil, err
+	}
+	// Validate the applyStrategy.
+	if !reflect.DeepEqual(updateRun.Status.ApplyStrategy, updateRunCopy.Status.ApplyStrategy) {
+		mismatchErr := fmt.Errorf("the applyStrategy in the clusterStagedUpdateRun is outdated, latest: %v, recorded: %v", updateRunCopy.Status.ApplyStrategy, updateRun.Status.ApplyStrategy)
```

Comment:
is this an user error?

Reply:
We could say updating the ApplyStrategy during a running stagedUpdateRun is a user error. Updated, thanks.